Blinkenlights

densminger opened the pull request:
Blinkenlights feature, as discussed in issue #766

densminger committed:
Added blinkenlights: and blinkenlight_player:

The purpose of blinkenlights are to allow multiple  modes to be able to control the same RGB light.
In this way, if multiple modes share the same shot, those modes can each add their own color to the
light for that shot, and the shot will flash all the different colors to let the player know that
the same shot will be scored for multiple modes. (For example, a multiball and a mode sharing the
same ramp shot.)
Modes can add colors to a blinkenlight with the blinkenlight_player, and the blinkenlight will
cycle through all the colors that have been added.  Modes can also remove colors, or remove all
colors, effectively turning the blinkenlight off.

densminger committed:
Added tests and fixed bugs to pass tests.

densminger committed:
Added .vscode to .gitignore

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,61 @@
+    machine_collection_name = 'blinkenlights'
+    allow_placeholders_in_keys = True
+
+    def __init__(self, machine):
```
This method is the same as the default. Could be removed.

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,61 @@
+            return
+        # only add this color if the key does not already exist
+        if len([x for x in blinkenlight.colors if x[1] == key]) < 1:
+            blinkenlight.colors.append((color, key))
```
I would turn this into a method in the blinkenlight device.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,80 @@
+        self._color_i = 0
+        self._num_colors = 0
+        # Add this device to the blinkenlight section
+        self.machine.blinkenlights[name] = self
```
This should not be needed here.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,80 @@
+
+    @property
+    def num_colors(self):
+        # self.info_log('num_colors queried - returning {}'.format(len(self._colors)))
```
Please remove commented out code.

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,61 @@
+            elif action == 'removeall':
+                self._remove_all_colors(blinkenlight)
+            blinkenlight.num_colors = len(blinkenlight.colors)
+            blinkenlight._restart()
```
Not sure if restarting this here is wise. It would destroy sync and might cause flickering on mode start/stop.

densminger replied:
I think this is ok, since now the blinkenlight figures out where it should be in the cycle, so it shouldn't interfere with the sync or cause flickering.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,80 @@
+        self._perform_step()
+
+    def _perform_step(self):
+        if len(self._colors) == 0:
```
I would turn this into a coroutine. We just need to figure out how we synchonize all blinkenlights. Maybe always start at sync_ms xx. Would that work? That would slightly change the logic. I looked at some sterns and it seems that they got up to 2 slots (i.e. on star trek). If both are filled the light will toggle between them. If one is filled it will turn off. Lights seem to be mostly in sync.

densminger replied:
I'll have to think about this.  Are you suggesting that each color change is in sync (so that each blinkenlight will change from one color to the next in sync with the others), or that each blinkenlight starts/finishes its entire cycle in sync will all other blinkenlights?  In the second case, I don't know how it would work if one blinkenlight had 2 colors and another has 6, since it would take the 6 color blinkenlight three times as long to cycle through its colors as the 2 color blinkenlight, so they couldn't stay in sync.

jabdoa2 replied:
I would imagine to sync the cycles. One idea would be to run them at a fixed duration and fill up the remaining slots with black (or transparency). This seems to be the case for sterns.

I see that this would be an issue with variable length cycles. Alternatively, we could only sync same length cycles (by using sync_ms = cycle length). That might still work well for the common case: green + off and green + red would sync but green + red + yellow would not sync. Guess that would be ok.

What do you think?

densminger replied:
I’m thinking I could make both options happen. Either you could specify a cycle_duration=1 for example, so that each cycle is 1 second regardless of how many colors there are, or color_duration=1 would specify that each color lasts 1 second. And then the sync_ms would sync accordingly (either on a cycle or color basis). Would that work?

atummons replied:
So with Stern, Game of Thrones has 3 colors that can be on a single light at a time.  I have a slow mo video, if you want.  I could upload and share the link.

The way it works is all are the same duration.  Lets say 1 for example.  Then each light has its cycle of what it can do.  LED1 has 1 color (R), LED2 has 2 colors (R,B), and LED3 has 3 colors (R,B,G).  They all loop through their colors and go to off at the end before restarting.  So here is a time breakdown.

Time 0: LED1 = R, LED2 = R, LED3 = R
Time 1: LED1 = Off, LED2 = B, LED3 = B
Time 2: LED1 = R, LED2 = Off, LED3 = G
Time 3: LED1= Off, LED2 = R, LED3 = Off

This would repeat.  So they aren't all showing red at once, or anything like that.  But that all shift at the exact same time.  So we would need a sync_ms for all blinkenlights (not each individual light), and a fixed duration for all lights (not each individual light).  This would mimic Stern to make it seamless to the user, as you have things all shifting simultaneously, instead of randomly.

I don't believe we need to enforce all of those rules, but rather expose the variables.  So I can set all of mine to duration 1 and fade_ms of 100.  Then mine will sync up.  But I would not set each to have 4 slots always, that are either filled or blank.  I would make them dynamic, so that they fill the slots that they have.  I had started code for this as well.  So I might play a bit more to see if it even works.  Then maybe we can compare and cherry pick best of class for inclusion in MPF.

jabdoa2 replied:
Thanks for checking that. We should support it this way.

Can you check one more thing for me: If you got two lights with R+B+off are those always in sync? I mean does both show red at the same time? I would guess so.

Basically, the above is what we should get with sync ms based on cycle length. Maybe, we can hack this together in the next few days.

atummons replied:
Yes, all binkenlights are exactly on the same sync.  So they all hit color 1 at the exact same time, color 2, etc.  So if two lights have the same values (R, B, Off), they would be perfectly in sync.  Then when you hit the R on shot 1, that gets removed from the loop for that light.  So then they aren't showing the same colors at the same time (1 showing B, Off and the other showing R, B, Off), but the sync_ms for transition is identical.  Here is a link to a video showing this on Game of Thrones: https://pindev.slack.com/files/U01GQ857T4L/F02A78ME7KR/img_8126.mov

densminger replied:
Thanks. I believe the way I've outlined would work then. What you're describing for how GoT works would be just like setting color_duration, so that each color on a particular light lasts for a given amount of time. The sync issue still might need some thought, since a sync_ms on each blinkenlight wouldn't ensure that they all started their cycle at the same time. For example, if I have 2 blinkenlights and they're each set with color_duration=500ms, and sync_ms=1s, this would ensure they started at the same time if they each had 2 colors. But if they each had 4 colors, they could be a half-cycle off from each other.

atummons replied:
That's a good point.  So might need to be at a higher level.  let me do some "research" on Game of Thrones.  I can get a mode started, and then add shots by progressing the modes.  I will see if it then syncs up with all shots or not.  I have a suspicious the answer will be yes.

But regardless of how Stern implemented, it would be nice to have them all start from the same place.  But with long duration, that becomes a bit of lag between when the event occurs and the lights add that color.

densminger replied:
How about this:

Blibkenlight_player figures out the sync_ms automatically, based on the current cycle length of the blinkenlight (I.e., the number of colors the light currently has if color_duration is set, or else the cycle_duration value if it is set). This is probably what Jan was suggesting earlier. Based on this, the blinkenlight_player could figure out which color should be on at any given time so that it wouldn't have to wait till the beginning of its cycle if the sync_ms happens to be very large when it starts.

In this case, there would be no user-configurable sync_ms option for blinkenlights, since it would be calculated and synced automatically.

However this does bring up another issue. This would require that all colors are added to the blinkenlight in the same order, or else some other ordering would be required. If one blinkenlight had its colors added R then G but reversed for the other blinkenlight, then these would flash alternating colors. 

(Also, this discussion should probably be moved to the issue #766 tracker instead of here on the pull request)

densminger replied:
In the latest commit, I addressed all these.  I added different parameters:

```
blinkenlights:
  my_blinkenlight:
    color_duration: 1s
    off_when_multiple: false
    light: l_stark_sigil
    priority: 1000
  my_blinkenlight2:
    cycle_duration: 1s
    off_when_multiple: false
    light: l_martell_sigil
    priority: 1000

```
Now you can either specify color_duration or cycle_duration (you have to specify one or the other, but not both - this is validated in code).  Sync across multiple blinkenlights is achieved based on the logic I outlined above (i.e., the cycle duration is automatically calculated and a "virtual" sync_ms is calculated based on the length of the cycle).  When the blinkenlight starts, it doesn't have to wait till the beginning of the next cycle (as it would with a typical sync_ms), but will instead figure out which color should be on at that point and go from there.

Also I added a priority.  This defaults to 0 if not specified.

jabdoa2 replied:
That sounds awesome!

> Sync across multiple blinkenlights is achieved based on the logic I outlined above (i.e., the cycle duration is automatically calculated and a "virtual" sync_ms is calculated based on the length of the cycle). When the blinkenlight starts, it doesn't have to wait till the beginning of the

We have to check how that looks. Might give us weird effects im some cases. One slightly more complex alternative would be to always finish the current cycle and only considering changes when the loop starts. Guess that would be something we might add later.

densminger replied:
Yes this could be weird in some cases but it works in a predictable manner. For example, if a blibkenlight has 2 colors (let's say red and green), so it blinks red, green, red, green. If green is removed by a mode while the green light is lit, then for the rest of green's "slot" time, it would the light would be off since now the blinkenlight should flash red, off, red, off. This could cause it to look like it flickers but it's intended behavior in this case.

jabdoa2 reviewed (commented):
I like it. Some small nits and some thoughts about sync.

densminger committed:
Changed parameters for blinkenlights.

densminger committed:
Added priority for blinkenlights.

densminger committed:
Added code to ensure blinkenlights are in sync with each other.

densminger committed:
Removed unneeded variable/method.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,131 @@
+
+    @property
+    def num_colors(self):
+        # I would prefer this to be a calculated value (i.e., just return len(self._colors)), but then there's no setter involved so things like mpf monitor won't know when this value gets updated.
```
There are some magic methods on DeviceMonitor to notify it about changes.

densminger replied:
I tried to find the magic methods you mentioned but I don't see any on DeviceMonitor aside from `__init__` and `__call__`.  I know `__setattr__` is used to notify subscribers when a variable is changed, but how can I notifiy subscribers without creating a fake variable like this?

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,131 @@
+    def num_colors_in_cycle(self):
+        # similar to num_colors, but adds 1 for the "off" color if there is one between cycles
+        return self.num_colors + (1 if self._off_between_cycles() else 0)
+
```
This additional newline probably makes the linter unhappy.

jabdoa2 commented:
Guess we are good to land this. Some small coding style changes to fix and we can merge this:

```
Messages

========

mpf/config_players/blinkenlight_player.py

  Line: 2

    pylint: unused-import / Unused deepcopy imported from copy

  Line: 4

    pylint: unused-import / Unused RGBColor imported from mpf.core.rgb_color

  Line: 5

    pylint: unused-import / Unused Util imported from mpf.core.utility_functions

  Line: 29

    pylint: protected-access / Access to a protected member _restart of a client class (col 12)

  Line: 31

    pylint: no-self-use / Method could be a function (col 4)

  Line: 36

    pylint: no-self-use / Method could be a function (col 4)

  Line: 41

    pylint: no-self-use / Method could be a function (col 4)

mpf/devices/blinkenlight.py

  Line: 3

    pylint: unused-import / Unused partial imported from functools

  Line: 7

    pylint: unused-import / Unused ColorException imported from mpf.core.rgb_color

  Line: 13

    pep8: E302 / expected 2 blank lines, found 1 (col 1)

    pep8: E302 / expected 2 blank lines, found 1 (col 1)

  Line: 30

    pep257: D102 / Missing docstring in public method

  Line: 33

    pylint: attribute-defined-outside-init / Attribute '_color_duration' defined outside __init__ (col 8)

  Line: 34

    pylint: attribute-defined-outside-init / Attribute '_cycle_duration' defined outside __init__ (col 8)

  Line: 35

    pep8: E501 / line too long (150 > 120 characters) (col 121)

  Line: 36

    pep8: E501 / line too long (126 > 120 characters) (col 121)

  Line: 38

    pyflakes: F821 / undefined name 'NoReturn' (col 66)

  Line: 42

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method

  Line: 43

    pep8: E501 / line too long (197 > 120 characters) (col 121)

  Line: 47

    pep8: N803 / argument name 'newValue' should be lowercase (col 27)

    pep8: N803 / argument name 'newValue' should be lowercase (col 27)

    pylint: invalid-name / Argument name "newValue" doesn't conform to '^[a-z_][a-z0-9]*((_[a-z0-9]+)*)?$' pattern (col 4)

  Line: 54

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method

  Line: 58

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method

  Line: 63

    pep8: E303 / too many blank lines (2) (col 5)

    pep8: E303 / too many blank lines (2) (col 5)

  Line: 72

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method

  Line: 79

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method

  Line: 84

    pylint: missing-function-docstring / Missing function or method docstring (col 4)

    pep257: D102 / Missing docstring in public method
```

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,48 @@
+                self._remove_color(blinkenlight, s['key'])
+            elif action == 'removeall':
+                self._remove_all_colors(blinkenlight)
+            blinkenlight._restart()
```
Guess this could be included in add/remove of the blinkenlight device.

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,48 @@
+        for blinkenlight, s in settings.items():
+            action = s['action']
+            if action == 'add':
+                self._add_color(blinkenlight, s['color'], s['key'])
```
We should pass priority here (`priority + s['priority']`).

densminger replied:
This makes me think. I believe that including priority here would include the mode's priority with the color.  This would mean that each color of a blinkenlight could have a different priority.  This seems strange to me.  The way I have it now is, the blinkenlight itself has a priority, and colors added to it all get the same priority regardless of the mode that added it.  That way, all of the blinkenlight's colors will either get shown or not shown as a group, instead of some colors being shown and others not shown due to the priority of the mode that added the color.

Another thing to think about is, what about the "off" color between cycles?  This isn't explicitly added by any mode, so what should its priority be?  If it's 0, then the "off" color might not get displayed at all which might look weird.

I'm not sure how this should be handled truthfully.

jabdoa2 replied:
Naming is a bit confusing here. `priority` on the light stack is one thing (which you already implemented). Another thing would be mode or stack priority or order. That would determine if red or blue comes first in your blinkenlight. This way the order should become deterministic (and based on mode priority). Makes sense?

densminger replied:
Got it. Thanks.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,131 @@
+
+    def add_color(self, color, key):
+        # only add this color if the key does not already exist
+        if len([x for x in self._colors if x[1] == key]) < 1:
```
Guess we should just remove the entry if it is already in. This allows updates.

jabdoa2 commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,131 @@
+            self._colors.append((color, key))
+            self.num_colors += 1
+            self.info_log('Color {} with key {} added'.format(color, key))
+
```
I would order the list by priority here

densminger replied:
I order the list in _restart, since add_color, remove_color, and remove_all_colors calls _restart.

atummons commented:
Looking good.  Well I haven't had time to look at mine any further, and you have already solved the problems, so I'll bow out here on fixes.  I think it is looking great with Jan's comments there.

Jan, how do we pull this down to run against our local machine?  I'd love to pull it down and test it out with Monitor to see how its works.  Or do we have to have it merged before that can be done?

jabdoa2 commented:
> Jan, how do we pull this down to run against our local machine? I'd love to pull it down and test it out with Monitor to see how its works. Or do we have to have it merged before that can be done?

I roughly explained this in my video on the IDE setup. Short:

- git clone https://github.com/densminger/mpf.git (clone Daves repo)
- cd mpf (change into new dir)
- git checkout blinkenlights (change to daves branch)
- pip3 install -e . (install in dev mode)

There are other ways with multiple remotes in one checkout but those are a bit tricky to use unless you are very familiar with the inner workings of git.

atummons commented:
> 
> 
> 
> I roughly explained this in my video on the IDE setup. Short:
> 
>     * git clone https://github.com/densminger/mpf.git (clone Daves repo)
> 
>     * cd mpf (change into new dir)
> 
>     * git checkout blinkenlights (change to daves branch)
> 
>     * pip3 install -e . (install in dev mode)
> 
> 
> There are other ways with multiple remotes in one checkout but those are a bit tricky to use unless you are very familiar with the inner workings of git.

Haven't watched that one yet (in my saved for later on YouTube), and I love the content!!!  I will check it out to make sure I understand the ins and outs.  But basically same as getting my local, just swapping the steps for Dave, and then do mine to switch back.  Makes perfect sense!

densminger commented:
So here's an example of some of the "weirdness" that might happen:

Let's say the blinkenlight has `cycle_duration=1s` (that is, each cycle lasts 1 second, regardless of how many colors the blinkenlight has).  The blinkenlight only has one color (red). In this case, the blinkenlight's cycle would look like this:

```
|       1s      |       1s      |       1s      |       1s      |
|-------|-------|-------|-------|-------|-------|-------|-------|
|  red  |  off  |  red  |  off  |  red  |  off  |  red  |  off  |
```

If green color is added to the blinkenlight (this assumes `off_when_multiple=false`), the cycle would look like this:

```
|       1s      |       1s      |       1s      |       1s      |
|-------|-------|-------|-------|-------|-------|-------|-------|
|  red  | green |  red  | green |  red  | green |  red  | green |
```

Then a third color (blue) is added:

```
|       1s      |       1s      |       1s      |       1s      |
|  r |  g  |  b |  r |  g  |  b |  r |  g  |  b |  r |  g  |  b |
```

Note that each color now is only 1/3 of a second long, since there are three of them per cycle now.

Now, let's say blue is removed from the blinkenlight, while the blinkenlight is currently showing a blue color:

```
|       1s      |       1s      |       1s      |       1s      |
|  r |  g  |  b |  r |  g  | b|g|  red  | green |  red  | green |
        blue is removed here ^
```

Notice how blue is displayed when the color is removed, and the light immediately switches to green, since green should be displayed at that point in time now that the blinkenlight only has 2 colors.  So the end result is green "flashes" very briefly before red is displayed again and the red/green cycle starts.

I hope this makes sense.  It's predictable, and I think if I was playing the game and saw this happen I would understand what happened and why the color change occurred like that.

jabdoa2 commented:
> So here's an example of some of the "weirdness" that might happen:
> 
> Let's say the blinkenlight has `cycle_duration=1s` (that is, each cycle lasts 1 second, regardless of how many colors the blinkenlight has). The blinkenlight only has one color (red). In this case, the blinkenlight's cycle would look like this:
> 
> ```
> |       1s      |       1s      |       1s      |       1s      |
> |-------|-------|-------|-------|-------|-------|-------|-------|
> |  red  |  off  |  red  |  off  |  red  |  off  |  red  |  off  |
> ```
> 
> If green color is added to the blinkenlight (this assumes `off_when_multiple=false`), the cycle would look like this:
> 
> ```
> |       1s      |       1s      |       1s      |       1s      |
> |-------|-------|-------|-------|-------|-------|-------|-------|
> |  red  | green |  red  | green |  red  | green |  red  | green |
> ```
> 
> Then a third color (blue) is added:
> 
> ```
> |       1s      |       1s      |       1s      |       1s      |
> |  r |  g  |  b |  r |  g  |  b |  r |  g  |  b |  r |  g  |  b |
> ```
> 
> Note that each color now is only 1/3 of a second long, since there are three of them per cycle now.
> 
> Now, let's say blue is removed from the blinkenlight, while the blinkenlight is currently showing a blue color:
> 
> ```
> |       1s      |       1s      |       1s      |       1s      |
> |  r |  g  |  b |  r |  g  | b|g|  red  | green |  red  | green |
>         blue is removed here ^
> ```
> 
> Notice how blue is displayed when the color is removed, and the light immediately switches to green, since green should be displayed at that point in time now that the blinkenlight only has 2 colors. So the end result is green "flashes" very briefly before red is displayed again and the red/green cycle starts.
> 
> I hope this makes sense. It's predictable, and I think if I was playing the game and saw this happen I would understand what happened and why the color change occurred like that.

Agreed. It definitely works this way :-). I guess we should add the ordering by priority in `blinkenlights_player` and with that we should be good to go.

densminger committed:
Syntax changes, and sorting colors by mode priority.

densminger commented:
I think this is ready to go.

atummons commented:
@densminger Looks like it needs a couple of cleanup items to pass the Travis CI build test.  All looks like general formatting and consistency stuff.  Once that's cleaned up, it should pass all three tests and get the green check mark.  Great work!
```
mpf/config_players/blinkenlight_player.py
  Line: 28
    pylint: missing-function-docstring / Missing function or method docstring (col 4)
    pep257: D102 / Missing docstring in public method
  Line: 34
    pylint: missing-function-docstring / Missing function or method docstring (col 4)
    pep257: D102 / Missing docstring in public method
  Line: 40
    pylint: missing-function-docstring / Missing function or method docstring (col 4)
    pep257: D102 / Missing docstring in public method

mpf/devices/blinkenlight.py
  Line: 15
    pylint: ungrouped-imports / Imports from package typing are not grouped (col 4)
  Line: 45
    pep8: E502 / the backslash is redundant between brackets (col 50)
    pep8: E502 / the backslash is redundant between brackets (col 50)
  Line: 53
    pep257: D401 / First line should be in imperative mood; try rephrasing (found 'Number')
  Line: 65
    pep257: D401 / First line should be in imperative mood; try rephrasing (found 'The')
  Line: 97
    pep257: D300 / Use """triple double quotes""" (found "-quotes)
  Line: 104
    pep257: D401 / First line should be in imperative mood (perhaps 'Remove', not 'Removes')
```

densminger committed:
Added some style changes for the linter.

densminger committed:
Removed unused import.

densminger committed:
Pylint fixes.

densminger committed:
Removed unused imports.

densminger commented:
Green check!

atummons commented:
Sweet!!!

densminger committed:
blinkenlights is only valid in machine

densminger committed:
sync_ms is no longer valid

densminger commented:
I've been playing with this, and I have a question.  Right now, a `blinkenlight` is a `SystemWideDevice` object.  As such, it exists at a machine level, and the `blinkenlight_player` has no concept of a game or player state.  So, when a mode adds a color to a blinkenlight, and the player's ball drains, that blinkenlight will continue to flash unless the mode specifically tells it to remove its colors when the mode ends (and it would have to "remember" which colors were active when it restarts).  Same thing if the game ends. My question is, is this ok, or should I change the `blinkenlight` to be a `LogicBlock` instead so it can use the `persist_state` functionality of logic blocks?  It's not exactly a logic block since it interfaces with hardware (lights) so I don't know if this is a good fit or not.

Edit: I just realized that the `ConfigPlayer` class has a `mode_stop` method.  I'm guessing I could override that method and unload the blinkenlight when modes end.  I would just need to keep track of which mode added which colors. So there's likely enough functionality already in `ConfigPlayer` to handle this correctly.

Edit2: I've marked this PR as draft.  I want to get this working before it gets merged.

atummons commented:
I was looking at the same thing when I saw your edit. Agreed, just need to use that to cancel when mode ends there and it clears the stack for it's called keys.

As for starting back up, is there a scenario where it would need to do that? I would assume my shot would be in a state that it's profile calls a specific show. So if the shot is in that state when the mode starts again, it should automatically be calling the show, and adding it back to the stack.

densminger commented:
Good point.  I think it's ok to not remember state.  I have it working so that when mode ends it removes the mode's colors automatically.  I changed the `removeall` action to `remove_all` (to make it consistent with other actions in mpf that use underscores) and I added another action called `remove_mode` that will remove all the colors added by the calling mode (just in case you want to remove all colors at a point other than when they're automatically removed at mode stop).

Thanks for your input, it helps a lot!

I'll push these changes shortly.

densminger committed:
Blinkenlight_player now removes all mode's colors when mode ends.

jabdoa2 commented:
> Edit: I just realized that the `ConfigPlayer` class has a `mode_stop` method. I'm guessing I could override that method and unload the blinkenlight when modes end. I would just need to keep track of which mode added which colors. So there's likely enough functionality already in `ConfigPlayer` to handle this correctly.

We already got a way to handle this. It is called "clear_context“. That is called when modes are unloaded, shows are stopped and in any other case. You can have a look at LightPlayer. It is similar to your approach but covers aboslutely everything in MPF. Also you probably do need the change in Blinkenlight with that.

densminger committed:
Changed the "key" parameter of blinkenlight_player to "label" to be clearer. Also fixed the failing tests.

densminger committed:
Fixes for Travis CI

densminger committed:
Travis CI fixes

densminger committed:
Changed references to "mode" to "context" instead to be more accurate.

densminger commented:
Got it. Thanks, Jan.  I changed it to use `clear_context` instead.  So blinkenlight colors are removed when the associated mode/context ends.  Example, mode1 adds blue and green, mode2 adds red.  When mode2 ends, red will automatically be removed from the blinkenlight, and when mode1 ends, blue and green will be removed.  However, a caveat to this is, the colors are also removed when a show ends.  So, if a show uses `blinkenlights:` to add a color to a blinkenlight, then the color will only flash while the show is active, and as soon as that show ends, the blinkenlight's colors will be removed.  I am guessing this is ok, just needs to be documented?  Let the user know that if they want a blinkenlight's color to remain active for a mode, the mode should add the color to the blinkenlight via a `blinkenlight_player` and not a through a show.  Speaking of which, I'm working on the documentation as well.

I also renamed the `key` parameter in `blinkenlight_player` to `label` instead, because I think that's clearer to the end-user as to its purpose.  Finally, I added an express config for `blinkenlight_player` which does the add action with a uuid/random label.

densminger committed:
No reason to store blinkenlights locally in blinkenlight_player.

jabdoa2 commented on mpf/config_players/blinkenlight_player.py:
```diff
@@ -0,0 +1,67 @@
+        """Adds/removes colors to blink."""
+        del kwargs
+
+        for blinkenlight, s in settings.items():
```
This is kind of different than most of our other players. Not sure if better or worse. One issue might be consistency.

Usually you can add and remove stuff with or without a `key`. Here this is called `label`. However, here you kind of have to use a label or removing your entry will not work later on (because you do not know the uuid). Also what you can do here is to remove `labels` from other modes. `keys` are namespaced in other players so that you cannot have conflicts between modes. Then we got contexts and labels here. In other places we just map remember our keys insides a context dict. Here `blinkenlight` needs to know both `context` and `label`.

Any thoughts on this? I'm always open to suggestions for improvement.

densminger replied:
So is your main concern that a mode can remove a color that was added by another mode if it uses the same label? My thought there is, that's kind of intended. Honestly, I think most of the time the user won't even use a `label` since the color is automatically removed when the mode ends. But it could be that you want to remove the color when a shot is hit or something.

jabdoa2 replied:
> So is your main concern that a mode can remove a color that was added by another mode if it uses the same label? My thought there is, that's kind of intended. Honestly, I think most of the time the user won't even use a `label` since the color is automatically removed when the mode ends. But it could be that you want to remove the color when a shot is hit or something.

I see that this might be a feature but it might as well confuse people. Conceptually, I think that it would be bad to design something which is added in one mode and removed in another one. That will lead to confusion and issues down the road.

In my opinion the main usecase would be this (which currently will not work this way):

```
blinkenlight_player:
  mode_my_mode_started:
    my_blinkenlight1: orange
  some_other_event:
    my_blinkenlight1: remove/stop    # some magic color. we use "stop" for lights
  some_other_event2:
    my_blinkenlight1:
      action: remove
```

I would simplify `blinkenlights` a bit to now know about contexts. Also I would use `key` instead of `label` for consistency with other players in MPF. If you want me to make those changes I can do that as well. We can do it on stream on Saturday if you like. Not 100% sure I will make it but I actually might.

atummons replied:
As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?

I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.

jabdoa2 replied:
That should already work. Your profile will stop and start shows. Stop will call clear_context and remove the previous color. Guess we should add a test with a show but I am quite confident that it will just work.

jabdoa2 replied:
Just to make this clear: I am very happy with the code how it is. I would like to change `label` to `key` for consistency and unexpose the context. Thats about it. Config interface is hard to change later. Refactoring the internal code can be done later and nobody will notice.

atummons replied:
Awesome. Well I can play around with it as I'm thinking and add tests as we go.
And as Jan said, awesome job! I think this will be a popular feature!

densminger replied:
Hmm. Something to consider if we go with what you're suggesting. What happens if a mode adds several colors (not sure why a mode would but it's totally possible and is designed to allow this). Then what would happen if:

```
blinkenlight_player:
  some_event:
    my_bl: remove
```

Would this remove all colors from the mode? Just the latest one?

Also, if you want to add or change any of this on stream, please feel free!

densminger replied:
> As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?
> 
> 
> 
> I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.

This wouldn't quite work because the blinkenlight will remove the show's colors when the show ends. So if your shot profile called a show which added a color, that color would only exist for the duration of the show.

The way you could do this now though, is to use blinkenlight_player to add a color when the shot is made:

```
blinkenlight_player:
  shot_left_ramp_hit:
    blinkenlight_left_ramp: red
```

atummons replied:
Yeah, that's what I wasn't sure about. If I had a show similar to on, then it would never end. So theoretically it would be in that state until I told it to end that show, which would remove it.

Otherwise that was my though, I add it when mode starts, and then xx_lit_hit would be my trigger to remove it. I'm excited to test this through the scenarios!

densminger replied:
> I add it when mode starts, and then xx_lit_hit would be my trigger to remove it. I'm excited to test this through the scenarios!

Yep. Just keep in mind that (as of now) if you want to remove the color yourself (as opposed to automatically removing the color when the mode ends) then you'll need to specify a `label` value when you add the color, then `remove` using that label later.

atummons replied:
Yep, and that's consistent with keys to remove widgets and other players. So people should be able to follow that logic through to here with a nice write up in the docs. I'm more than willing to assist with that piece for nice we get the code finalized.

densminger replied:
You can check out my PR on the mpf-docs repo. I've already added documentation for blinkenlight in that pull request. I'm getting a Travis CI error though and I'm not sure why.

jabdoa2 replied:
> Hmm. Something to consider if we go with what you're suggesting. What happens if a mode adds several colors (not sure why a mode would but it's totally possible and is designed to allow this). Then what would happen if:
> 
> ```
> blinkenlight_player:
>   some_event:
>     my_bl: remove
> ```
> 
> Would this remove all colors from the mode? Just the latest one?

In this case you should use keys. Otherwise the second color should remove the first one (because they got the same key). That is the logic for all other players at least. Except maybe in widget player where you can intentionally leak objects to add one widget multiple times (prone to memory leaks obviously).

> > As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?
> > I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.
> 
> This wouldn't quite work because the blinkenlight will remove the show's colors when the show ends. So if your shot profile called a show which added a color, that color would only exist for the duration of the show.

Basically, your color has to be included in all shows where it should exist. Similar to light_player behaviour.

atummons replied:
Played around with this and so far with basic tests its working great.  I tried adding it to shot_profiles, and it worked just as Jan said it would.  When I dug in, I did run into one error though.  Placeholders for blinkenlights work fine, but for color threw an error.
```
shows:
    bl_on:
      - duration: -1
        blinkenlights:
            (light): (color_1)

File "c:\users\atummons\desktop\davebranch\mpf\mpf\core\config_validator.py", line 683, in _validate_type_color return int(color[0]), int(color[1]), int(color[2])
ValueError: invalid literal for int() with base 10: '(color_1)'
```
Not sure why this works for shows with lights, but not blinkenlights.  I looked around, but don't really know what I was looking for.

jabdoa2 replied:
Good find. I think we need color_or_token as type to support that.

atummons replied:
I made the comment in the config spec.  Didn't need to make it a token.  But in config spec it was a type color, but needs to be a type str.  This is the same as light_player.  I made the change local, and it worked as expected.  So I made a comment on that row for the change.

Not sure what's appropriate and good coding practices.  Should I make a push for that change to someone else's branch?  Or just note it for them?  Not sure what standard practice is, or if its all personal preference for Dave, since this is his branch.

densminger replied:
Go ahead and make the change if you want. I was holding off on making changes in case Jan decided to do a stream today, but generally I have no problems if you want to make changes.

jabdoa2 replied:
Maybe we can finish this baby together later on. Should be fun.

atummons replied:
Its looking super awesome.  I put it through a few more paces this morning, and worked just as expected!  So I think a few of those cleanup things from this thread, and if we want to implement the color_or_token method and we should be in business!

densminger replied:
Woohoo!

atummons replied:
I put together a fix for a method call that was wrong, and added color_or_token, but saying I'm getting permission denied to push to Dave's branch.  So is best option to just post changes here and someone with authority can make them?  Or is there a way I can push against it and I'm trying to push to Dave's branch, instead of this PR?

densminger replied:
> I put together a fix for a method call that was wrong, and added color_or_token, but saying I'm getting permission denied to push to Dave's branch.  So is best option to just post changes here and someone with authority can make them?  Or is there a way I can push against it and I'm trying to push to Dave's branch, instead of this PR?

I have it set to allow changes from mpf maintainers (if you have write access to mpf you can change the PR) but I don't think I can allow changes from anyone else? My GitHub knowledge is lacking here lol

atummons replied:
Yeah, I'm not an mpf maintainer, I don't believe.  I have your branch checked out on local, and am pushing to your local.  I think you have to add me to your repo as a collaborator (don't do this, I have limited knowledge as well and am not to be trusted on your branches!).  So I can shoot it over (its only about 10 lines) if we get together to discuss final stuff.

densminger replied:
> I would like to change `label` to `key` for consistency and unexpose the context.

I renamed `label` back to `key`.  As far as "unexposed the context" I'm not sure what you would like to do there, exactly.  So I'll let you make those changes, if that's ok.

atummons replied:
Pulled this in and all is working as expected with both of your updates.

atummons commented on mpf/devices/blinkenlight.py:
```diff
@@ -0,0 +1,164 @@
+        self.info_log('All colors removed')
+        self._restart()
+
+    def remove_color_with_label(self, label):
```
This needs changed to private _remove_color_with_label (or if it needs to be public, line 92 needs changed to remove the underscore.

atummons replied:
Blinkenlight_player calls this, so it looks like the call on 92 needs to be global.

densminger replied:
Good catch.  That would have been a runtime error.  Made me realize I didn't have a unit test for this.  I've fixed it and added a test.  Will push in next update.

atummons commented on mpf/config_spec.yaml:
```diff
@@ -331,6 +339,12 @@ fast_coils:
+    __type__: config_player
+    action: single|enum(add,remove,remove_mode,remove_all)|add
+    label: single|str|None
+    color: single|color|ffffff
```
This should be changed to:     color: single|str|white
This will allow tokens to be passed in, in addition to color names of color codes.

jabdoa2 replied:
No. This would change to `single|color_or_token|white`

jabdoa2 replied:
Looks like you are right. This does not yet exist. I guess we should create it (as we have it for floats or ints). That will move str parsing out of runtime in most cases.

atummons replied:
Can't publish changes, but here's what they are.

Update 347 to:     color: single|color_or_token|ffffff.

Then in config_validator:
Add line 81:             "color_or_token": self._validate_type_or_token(self._validate_type_color),
Line 669: Swap above the two lines of comments (#) for the following:
    def _validate_type_color(self, item, validation_failure_info, param=None):
        assert not param
        if item is not None:
            try:
                if isinstance(item, tuple):
                    if len(item) != 3:
                        self.validation_error(item, validation_failure_info, "Color needs three components")
                    return item
            except ValueError:
                self.validation_error(item, validation_failure_info, "Cannot convert value to color.", 11)

densminger replied:
What's the purpose of `param=None` and `assert not param`?  These aren't in my version of config_validator and I'm not sure the purpose.  I'm not sure what's going on here in general.

atummons replied:
Jan would have to answer the specifics for you here. But we switched to allowing tokens, and it runs through the token validated to see is it a token on a color, in this case. If it's a token, we know that, so we stop before the validation breaks. If it's not a token, we go through. Without this argument it breaks here as it expects three arguments, but gets 4 due to the token validation.

This is based on the other token_or_xx methods here.

densminger replied:
Ah ok. I'll take a look when I get home and push the changes later tonight.

densminger committed:
Blinkenlight shows now support tokens for blinkenlight name and color. Also, fixed a bug if a blinkenlight color was added that already exists.  Also added tests for both of these.

densminger commented:
@atummons Ok, both of your fixes have been pushed.  I changed the `_validate_type_color` method a little from what you suggested since it didn't require everything you said.  I added tests for both the bug/typo you found, as well as the tokens, so these things should continue to work in the future.

densminger committed:
Renamed label parameter back to key (as per Jan's suggestion)

densminger commented:
Renamed `label` back to `key` as per Jan's suggestion. @atummons, this will mean you'll need to change your code that you're using for testing as well, just fyi.

atummons commented:
Awesome work! I'll get latest on your repo and continue to play around with my test machine.

jabdoa2 committed:
simplify code a bit and make key work as in other players

jabdoa2 committed:
properly cleanup at the end

densminger reviewed (commented):
Looks great! Everything works on my end.  It's ready to go as far as I'm concerned.

Edit: So it looks like the "stop/remove" express config option will only remove colors that were added without a key (that is, colors that were added with the express config method, or colors that were added without specifying a key).  This makes sense, I just need to update the documentation to reflect that.

jabdoa2 commented:
> Looks great! Everything works on my end. It's ready to go as far as I'm concerned.

Thanks for testing and confirming. This has been an awesome collaboration! Thanks @densminger @atummons .



> Edit: So it looks like the "stop/remove" express config option will only remove colors that were added without a key (that is, colors that were added with the express config method, or colors that were added without specifying a key). This makes sense, I just need to update the documentation to reflect that.

Yeah exactly. My theory is that 99% of the users will have maximum one color per mode.
